Enhancement/577 Threshold values for Embeddings

faisal-alvi opened the pull request:
<!--
Filling out this template is required.  Any PR that does not include enough information to be reviewed may be closed at a maintainers' discretion.  All new code requires documentation and tests to ensure against regressions.
-->

### Description of the Change

* This PR adds the threshold setting fields for each taxonomy in the Embeddings settings, allowing users to set the threshold value.
* The default value is set to 75%.
* These calculations are based on https://github.com/10up/classifai/issues/577#issuecomment-1808636943.

<!-- Enter any applicable Issue number(s) here that will be closed/resolved by this PR. -->
Closes #577

### How to test the Change
<!-- Please provide steps on how to test or validate that the change in this PR works as described. -->

### Changelog Entry
> Added - Threshold values for Embeddings


### Credits
<!-- Please list any and all contributors on this PR so that they can be added to this projects CREDITS.md file. -->
Props @timatron @dkotter @faisal-alvi 


### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you are unsure about any of these, please ask for clarification.  We are here to help! -->
- [ ] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests pass.

faisal-alvi committed:
render threshold fields

faisal-alvi committed:
sanitize the threshold settings

faisal-alvi committed:
pass the threshold settings

faisal-alvi committed:
fetch the terms in the cap of set threshold

faisal-alvi committed:
add `classifai_threshold` filter

faisal-alvi committed:
handle the threshold input

faisal-alvi committed:
clean the fields layout

faisal-alvi committed:
update the e2e test

faisal-alvi committed:
default threshold to 75 for existing users

dkotter commented on includes/Classifai/Providers/OpenAI/EmbeddingCalculations.php:
```diff
@@ -56,8 +58,15 @@ function( $x ) {
+		 * Filter the threshold for the similarity calculation.
+		 *
+		 * @param float $threshold The threshold to use.
+		 */
```
This hook docblock will need updated to match our other hooks so it ends up in our doc site

dkotter commented on includes/Classifai/Providers/Provider.php:
```diff
@@ -446,6 +453,34 @@ public function render_checkbox_group( array $args = array() ) {
+				<input type="number" id="%1$s_%2$s_%3$s" class="small-text" name="classifai_%1$s[%2$s][%3$s]" value="%5$s" />				
+			</p>',
+			esc_attr( $this->option_name ),
+			esc_attr( $args['label_for'] ),
```
Wondering if we need a simple check here to ensure this value actually exists? Something like

```
esc_attr( $args['label_for'] ?? '' )
```

dkotter commented on includes/Classifai/Providers/OpenAI/Embeddings.php:
```diff
@@ -354,6 +363,34 @@ public function supported_post_types() {
+		 *
+		 * @param float $threshold The threshold to use.
+		 * @param string $taxonomy The taxonomy to get the threshold for.
+		 */
```
Same comment on this hook docblock

dkotter reviewed (changes requested):
Note that there's some PHPCS warnings that we should look to fix: https://github.com/10up/classifai/actions/runs/6968812438/job/18963465180#step:8:16. Not sure why it's not showing this check as having failed, maybe we only fail on errors and not warnings? But these are easy enough to fix up so ideally we address these

EDIT: looked into this and seems it's due to the 10up-Default ruleset we extend, based on this line: https://github.com/10up/phpcs-composer/blob/master/10up-Default/ruleset.xml#L33

Personally I think we should remove this so we get failures on warnings, otherwise those items never get fixed up.

faisal-alvi committed:
hook docblock updates

faisal-alvi committed:
ensure value actually exists

faisal-alvi committed:
phpcs minor fixes

faisal-alvi commented:
@dkotter thanks for the feedback. I've implemented them (and also fixed the minor phpcs warnings). Please check.

> Personally I think we should remove this so we get failures on warnings, otherwise those items never get fixed up.

Agree. Should I raise a PR there?

dkotter commented on includes/Classifai/Providers/OpenAI/EmbeddingCalculations.php:
```diff
@@ -56,8 +58,20 @@ function( $x ) {
+		/**
+		 * Filter the threshold for the similarity calculation.
+		 *
+		 * @since 2.6.0
```
```suggestion
		 * @since 2.5.0
```

dkotter commented on includes/Classifai/Providers/OpenAI/Embeddings.php:
```diff
@@ -354,6 +363,39 @@ public function supported_post_types() {
+	/**
+	 * Get the threshold for the similarity calculation.
+	 *
+	 * @since 2.6.0
```
```suggestion
	 * @since 2.5.0
```

dkotter commented on includes/Classifai/Providers/OpenAI/Embeddings.php:
```diff
@@ -354,6 +363,39 @@ public function supported_post_types() {
+		/**
+		 * Filter the threshold for the similarity calculation.
+		 *
+		 * @since 2.6.0
```
```suggestion
		 * @since 2.5.0
```

dkotter commented on includes/Classifai/Providers/Provider.php:
```diff
@@ -446,6 +453,34 @@ public function render_checkbox_group( array $args = array() ) {
+	/**
+	 * Render a threshold field.
+	 *
+	 * @since 2.6.0
```
```suggestion
	 * @since 2.5.0
```

dkotter committed:
Update since version

dkotter committed:
Merge branch 'develop' into enhancement/embeddings

dkotter reviewed (changes requested):
In testing, found a few things:

1. When first going to the settings page, I don't see any default values for the thresholds. After saving the first time, they get set to 75% but those should be set on initial load:
![Default values blank](https://github.com/10up/classifai/assets/916738/a5236a83-370b-4619-ba0d-ca3e77bf6026)
2. If the Watson feature is turned off, doesn't seem we load the language-processing.css file, so styles aren't getting loaded:
![Missing styles](https://github.com/10up/classifai/assets/916738/6ad6c68e-0fc8-45dc-846d-f50015523cce)
3. Don't think this is a side-effect of this PR (probably due to changes in #609) but no longer seeing the Classify button if the feature is disabled on the post level:

![Missing button](https://github.com/10up/classifai/assets/916738/38699e6c-c1ad-460f-8e6f-c6bdd4e4c0c4)

dkotter commented:
> Agree. Should I raise a PR there?

Yeah, or just make that change in this PR, either way works for me

faisal-alvi committed:
set default to 75 for fresh setups

faisal-alvi committed:
load style for all tabs

faisal-alvi committed:
improve the condition to load style in editor

faisal-alvi committed:
ignore check for nonce as does not seem require

faisal-alvi committed:
phpcs: For CI, also fail on warnings

faisal-alvi commented:
Thanks for the feedback. Changes are implemented.

> Don't think this is a side-effect of this PR (probably due to changes in https://github.com/10up/classifai/pull/609) but no longer seeing the Classify button if the feature is disabled on the post level

I have found that this is because the post status is being `auto-draft` on the first load and the plugin can not find it in the allowed post types list.

![image](https://github.com/10up/classifai/assets/25176325/474eadd4-6850-4ecb-9338-a12938e37219)

So as a fix, we can allow `auto-draft` if someone has allowed the `draft` type; no need to add a new post type in the settings.

dkotter commented on includes/Classifai/Providers/Watson/NLU.php:
```diff
@@ -145,7 +145,7 @@ public function get_default_settings() {
 	 */
 	public function register() {
-		if ( $this->has_access( 'content_classification' ) ) {
+		if ( $this->has_access( 'content_classification' ) || isset( $_GET['page'] ) && 'classifai' === $_GET['page'] ) { // phpcs:ignore WordPress.Security.NonceVerification.Recommended
```
I'm not sure we'll want to add this conditional here. What this means is that even if the NLU feature is off, we register the NLU taxonomies anytime we're on any ClassifAI settings page (as well as load other functionality that isn't needed).

Unless there's something I'm missing, I think we can just take the same `enqueue_admin_assets` method here and load the correct assets in the Embeddings class, to ensure the assets we need are always loaded on that page. If both features are turned on, that's fine, as WordPress can handle the situation where an asset is enqueued twice.

dkotter commented:
> I have found that this is because the post status is being auto-draft on the first load and the plugin can not find it in the allowed post types list.

I was testing with an already published item so I don't think that's the issue I was facing. In looking closer, it seems we have a conditional in our `gutenberg-plugin.js` file that only loads that button if NLU is enabled. Seems it was set up this way from the beginning and so that button has never shown for the Embeddings feature (which I was mistaken there, I thought I remember it being there but I guess not). I think it would be nice to show that button but that doesn't need to be handled in this PR.

faisal-alvi committed:
enque assets for embeddings

faisal-alvi committed:
Merge branch 'develop' into enhancement/embeddings
